vs2010: fix target_to_build_root method

nioncode opened the pull request:
Python's os.path.split() does not split the path into its components.
Instead, split the path with str.split() using the OS's file system
separator.

I also added a test case that failed before and now passes, because the subdir folder is three levels deep.

jpakkane commented:
On Windows both `/` and `\` can be path separators and presumably (no access to Win atm so speculating)  `os.sep` is `\`. This would fail if paths with slashes sneak in, like for example when doing `files('foo/bar.c')`. Also, Meson internally tries to stick with `/` always. What is it specifically that breaks with `os.path.split` in this case?

nioncode commented:
`os.path.split` simply does not split a path like 'subprojects\sub\lib' into `['subprojects', 'sub', 'lib']`, but into `['subprojects\\sub', 'lib']`, see: https://docs.python.org/3/library/os.path.html#os.path.split.
The current code did then just count the elements in the array and created as many `../` as there were elements, assuming the number of elements are the number of directories, which is wrong and leads to `'../..'`, where it should be `'../../..'`, because the subdir path is 3 levels deep.

My patch uses `str.split` instead of `os.path.split`, which splits the path into all components as expected.

The method where I did apply the fix is only called to generate a relative path from the current target vcxproj to the build directory. It has nothing to do with source files, so it does not break anything here.

Regarding `/` vs `\`:
All paths (at least in the vs2010 backend) are created with `os.path.join`, so it is not possible for `/` to sneak in. However, it would be fine to call `os.path.normpath` first (this replaces all `/` with `\` on Windows), then do the path splitting with `os.sep` and finally join the path again with `/` instead of `os.sep` if you'd rather have `../../..` instead of `..\..\..`. In my opinion, though, meson should just use the native path separator on each OS.

So, should I refactor to use `/` instead of `os.sep` and take care of possible `/` before splitting the path?

jpakkane commented:
I'd say the easiest and most reliable approach is to call normpath and then split via os.pathsep.

nioncode commented:
Done :)
